chem/KPP/util/write_decomp/integr_edit.c

```diff
@@ -25,99 +25,84 @@ main( int argc, char *argv[] )
-  sprintf ( tfname, "%s_new", intfname );
 
-  
+  argv++;
+  strcpy( intfname, *argv );
+  argv++;
+  strcpy( incfname, *argv );
+  argv++;
+  strcpy( tfname, *argv );
 
-  sprintf(  cp_command,"cp %s %s",tfname, intfname);
+  sprintf( cp_command,"cp %s %s",tfname, intfname );
 
-  intf = fopen( intfname , "r" );
-  incf = fopen( incfname , "r" );
-  ofile = fopen( tfname , "w" );
+  intf  = fopen( intfname, "r" );
+  incf  = fopen( incfname, "r" );
+  ofile = fopen( tfname,   "w" );
 
 
   
-  sprintf ( callln , "   CALL %s_KppDecomp\0", mechname );
-  sprintf ( endln , "END MODULE");
-
-     /* loop over lines in Integr file */
-       while ( fgets ( inln , 4096 , intf ) != NULL ){
-
-	   copyit=1;
-      
-        
-	   /* replace call to decomp routine */
-
-           if ( !strncmp (inln, callln, strlen(callln)-1) ) {
-
-	  printf("   integr_edit: replacing  %s \n", inln);
-
-          fprintf(ofile, "!!!  use direct adressing in decomp \n");
-          fprintf(ofile, "!!! %s", inln);
-          fprintf(ofile, "CALL decomp_%s ( A, ising )\n", mechname );
+  sprintf( callln, "   CALL %s_KppDecomp\0", mechname );
+  sprintf( endln, "END MODULE" );
 
+  /* loop over lines in Integr file */
+  while ( fgets( inln , 4096 , intf ) != NULL ) {
 
-          add_sub=1;
-	  copyit=0;
-           }
+    copyit = 1;
 
 
-	   /* add decomp routine w. direct referncing */
+    /* replace call to decomp routine */
+    if ( !strncmp( inln, callln, strlen( callln ) - 1 ) ) {
 
-          if ( !strncmp (inln, endln, strlen(endln)-1) ) {
+      printf("   integr_edit: replacing  %s \n", inln);
+      fprintf(ofile, "!!!  use direct adressing in decomp \n");
+      fprintf(ofile, "!!! %s", inln);
+      fprintf(ofile, "CALL decomp_%s ( A, ising )\n", mechname );
 
-	    if ( add_sub ){
-	    printf("  %s ", inln );
-               while ( fgets ( incln , 4096 , incf ) != NULL ){
-
-                fprintf(ofile, "%s", incln);
-
-               }
-            
-                fprintf(ofile, " \n\n\n");
-            }
-          }
-
-
-
-
-       /* copy line from original file */
-	 if ( copyit ) {
- 
-	   fprintf(ofile, "%s", inln);
+      add_sub = 1;
+      copyit  = 0;
+    }
 
-	 } 
 
+    /* add decomp routine w. direct referncing */
+    if ( !strncmp (inln, endln, strlen(endln)-1) ) {
 
-     }
+      if ( add_sub ) {
+        printf("  %s ", inln );
+        while ( fgets ( incln , 4096 , incf ) != NULL ) {
+          fprintf( ofile, "%s", incln );
+        }
+        fprintf(ofile, " \n\n\n");
+      }
+    }
 
 
-       if ( ! add_sub  ) {
-	 printf(" integr_edit: Kept previous version. \n "); 
-       }
+    /* copy line from original file */
+    if ( copyit ) {
+      fprintf( ofile, "%s", inln );
+    }
+  }
 
+  if ( ! add_sub  ) {
+    printf( " integr_edit: Kept previous version. \n " );
+  }
 
 
   fclose( intf );
   fclose( incf );
   fclose( ofile );
 
-
-   system(cp_command);
+#ifndef NO_COPY
+  system(cp_command);
```

Comment:
Calling system() with a command-line-provided argument introduces significant security risks, as it can allow command injection if the input is not carefully sanitized. Could we replace system(cp_command) with a safer alternative that avoids directly invoking the shell?

For example, the following implementation securely copies files using fork() and execlp():
```
#include <unistd.h>
#include <sys/wait.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void copy_file(const char *source, const char *destination) {
    pid_t pid = fork();
    if (pid == 0) {
        // Child process: Securely execute the `cp` command
        execlp("cp", "cp", source, destination, NULL);
        perror("execlp failed"); // If exec fails
        exit(EXIT_FAILURE);
    } else if (pid > 0) {
        // Parent process: Wait for the child to complete
        int status;
        waitpid(pid, &status, 0);
        if (WIFEXITED(status)) {
            printf("Copy completed with exit code %d\n", WEXITSTATUS(status));
        } else {
            fprintf(stderr, "Copy process terminated abnormally\n");
        }
    } else {
        perror("fork failed");
        exit(EXIT_FAILURE);
    }
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s \"<source_file> <destination_file>\"\n", argv[0]);
        return EXIT_FAILURE;
    }

    // Parse the cp_command argument into source and destination
    char *cp_command = argv[1];
    char *source = strtok(cp_command, " ");
    char *destination = strtok(NULL, " ");

    if (!source || !destination) {
        fprintf(stderr, "Error: Invalid command. Provide source and destination files.\n");
        return EXIT_FAILURE;
    }

    // Call the copy_file function
    printf("Copying '%s' to '%s'...\n", source, destination);
    copy_file(source, destination);

    return EXIT_SUCCESS;
}
```